pip/pip-269.md

```diff
@@ -0,0 +1,242 @@
+
+In the Multi Consumer Dispatcher, there has a mechanism that tries to fix the issue caused by the action `cursor.rewind` and **Sequential Read** executed concurrently(referred to as Rewind After In-flight Reading). It works like this: if there has an in-flight reading when calling `cursor.rewind`, just do rewind after the in-flight reading is finished.` but this mechanism has the following drawbacks:
+- Makes a new issue lead to a new issue<sup>[5]</sup>.
+- Only applicable to `cursor.rewind`.
+- Only applicable to Multi Consumer Dispatcher.
+- If there is more than one in-flight reading(one Sequential Read and some Message Replay Read might exist simultaneously), this mechanism can not work.
+
+# Goals
+**Sub-Goal-1**: when a **Cursor Position Reset*** occurs, reject new reading, discard the in-flight reading and half in-flight reading<subp>[6]</sup>. And take care of these two things:
+- Ensure that changes to `read position` and `mark deleted position` are thread-safe.
+- Since more than one component uses Cursor to do **Sequential Read** or **Message Replay Read**, these components may also have their own attributes that need to be modified atomicity, we should also provide a mechanism to guarantee the atomicity modification of both attributes of Cursor and attributes of other components. 
+**Sub-Goal-2**: Remove some patchs(referred to as **Goal-3**):
+  - Revert https://github.com/apache/pulsar/issues/6403 (it named "A compensation mechanism of **Message Replay Read**" above)
+  - Remove the mechanism Rewind After In-flight Reading<sup>[5]</sup>
+**Out of scope**: Prevent more than one **Cursor Position Reset** executed at the same time(Current Proposal does not consider this optimization, there will be some bug fixes after this PIP)
+
+# High-Level Design
+
+### New concepts
+- **Cursor position is resetting**: indicate if there is a **Cursor Position Reset** running. Inject new readingg and discard in-flight reading if it is `true`.
+- **Cursor Epoch**: a counter to record how many times the the **Cursor Position Reset** executed. If it is increased after a in-flight reading, discard this reading.
+  - This attribute can help to prevent such error: a reading is started(`${Cursor position is resetting}` is `false`); a **Cursor position is resetting** is executed; the reading is completed(`${Cursor position is resetting}` is `false`).
+
+### Guarantee safe changes
+- Inside Cursor: the changes of attributes `read position`, `mark deleted position`, `cursor epoch`, and `cursor position is resetting` use the same lock.
+- Outside Cursor
+  - The action **Cursor Position Reset*** update `${Cursor position is resetting}` first, which guarantees that no other **Cursor Position Reset*** events will reenter(fast fail), and new reading can be rejected (In other words, it is the same as a non-reentrant lock).
+  - Each component (such as Dispatcher, Replicator...) modifies its attributes(each component to decide whether it needs to a locked).
```

Comment:
No, in this design it will work like this:

```java
public class Dispatcher {
 
  public void action1() {
     cursor. startRevision(); // runs in cursor lock block, release the cursor lock after this method completes
     lock.lock();
     clearRedeliverMessages();
     ....
     lock.unlock();
     cursor. endRevision( cursor -> cursor.rewind() ); // runs in cursor lock block, release the cursor lock after this method completes
  }

}

```